numpy/random/bit_generator.pyi

```diff
@@ -0,0 +1,126 @@
+import abc
+from threading import Lock
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Literal,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from numpy import dtype, ndarray, uint32, uint64, unsignedinteger
+from numpy.typing import (
+    DTypeLike,
+    _ArrayLikeInt_co,
+    _DTypeLikeUInt,
+    _ShapeLike,
+    _SupportsDType,
+)
+
+_T = TypeVar("_T")
+
+_UIntType = TypeVar("_UIntType", uint64, uint32)
+_DTypeLike = Union[
+    Type[_UIntType],
+    dtype[_UIntType],
+    _SupportsDType[dtype[_UIntType]],
+]
+
+class _Interface(NamedTuple):
+    state_address: Any
+    state: Any
+    next_uint64: Any
+    next_uint32: Any
+    next_double: Any
+    bit_generator: Any
+
+class ISeedSequence(abc.ABC):
+    @overload
+    @abc.abstractmethod
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLike[_UIntType] = ...
+    ) -> ndarray[Any, dtype[_UIntType]]: ...
+    @overload
+    @abc.abstractmethod
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLikeUInt = ...
+    ) -> ndarray[Any, dtype[unsignedinteger[Any]]]: ...
+
+class ISpawnableSeedSequence(ISeedSequence):
+    @abc.abstractmethod
+    def spawn(self: _T, n_children: int) -> List[_T]: ...
+
+class SeedlessSeedSequence(ISpawnableSeedSequence):
+    @overload
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLike[_UIntType] = ...
+    ) -> ndarray[Any, dtype[_UIntType]]: ...
+    @overload
+    def generate_state(
+        self, n_words: int, dtype: _DTypeLikeUInt = ...
+    ) -> ndarray[Any, dtype[unsignedinteger[Any]]]: ...
+    def spawn(self: _T, n_children: int) -> List[_T]: ...
+
+class SeedSequence(ISpawnableSeedSequence):
+    entropy: Union[None, int, Sequence[int]]
+    spawn_key: Tuple[int, ...]
+    pool_size: int
+    n_children_spawned: int
+    pool: ndarray[Any, dtype[uint32]]
+    def __init__(
+        self,
+        entropy: Union[None, int, Sequence[int]] = ...,
+        *,
+        spawn_key: Tuple[int, ...] = ...,
+        pool_size: int = ...,
+        n_children_spawned: int = ...,
+    ) -> None: ...
+    def __repr__(self) -> str: ...
+    @property
+    def state(
+        self,
+    ) -> Dict[str, Union[None, Sequence[int], int, Tuple[int, ...]]]: ...
```

Comment:
Continuing from https://github.com/numpy/numpy/pull/18395/files#r574883582 it is also possible to use `TypedDict` here.

``` python
import sys
if sys.version_info >= (3, 8):
    from typing import TypedDict, Literal
else:
    from typing_extensions import TypedDict, Literal

class _SeedDictBase(TypedDict):
    spawn_key: Tuple[int, ...]
    pool_size: int
    n_children_spawned: int

# A dict that contains the `spawn_key`, `pool_size` and `n_children_spawned`
# keys and, optionally, the `entropy` key
# The `total=False` keyword ensures that the `entropy` key is optional
class _SeedDict(_SeedDictBase, total=False):
    entropy: Union[int, Sequence[int]]

class SeedSequence:
    @property
    def state(self) -> _SeedDict: ...
    @state.setter
    def state(self, value: _SeedDict) -> None: ...
```

Reply:
Thanks.  Had to simplify and use a more basic definition.

Reply:
This comment still has to-be addressed.
Considering `SeedSequence` is not an abstractbaseclass I'd go with either the `TypedDict` approach or `Dict[str, Any]`.

Reply:
I've made it a `TypedDict`.